### ISSUE TITLE
Proposal: make `if` with a single statement that uses `defer` a compiler error 

### ISSUE THREAD

**Issue by Jarred-Sumner** [ziglang/zig#18911]
The following is an easy mistake for developers new to Zig:
```zig
if (condition) {
  // this runs immediately! not after now()
  defer later();
}

now();
```

Someone new to Zig would think "okay, `later()` is going to run after `now()`.", but actually it runs immediately.

Is there ever a case where a single-statement `if` with `defer` is not a bug? 

If there are no cases, could this be a compiler error?

For example, I am _not_ suggesting the following:
```zig
// these are all fine! 
if (condition) {
  defer later();
  anotherFn();
}
if (condition) {
  anotherFn();
  // should this be fine? i'm not sure
  defer later();
}
defer if (true) later();
defer {
  if (true) later();
 }
defer {
    if (true) {
      later();
    }
 }
```

Motivating example: https://github.com/oven-sh/bun/blob/f1d2fcca16f31b612e944ebe8f7de7abc8c94347/src/bun.js/javascript.zig#L2768-L2770

I am less certain of the behavior with `errdefer`

**Comment by rohlem** [comment]
I agree that a block/body with only `defer` will generally be an error, however, commenting-out code can also turn valid and intentional code into this shape.

```zig
if (condition) {
  defer a(n);
  b(&n);
}
// comment out b, would now be a compile error
if (condition) {
  defer a(n);
  //b(&n);
}
// to make it compile again
if (condition) {
  a(n); //defer when we reactivate a statement below
  //b(&n);
}
```

This is a bit similar to some of the `_ = &x;` statements that need to be introduced for unused/unmodified variable errors,
so there's definitely precedent to go ahead with it regardless (and I'm not personally opposed either).

I'm wondering whether it would make sense to additionally introduce a sort of "void statement"
to specifically silence / work around this new "immediate-defer" compile error.
Could be `_ = _;`, or even a named builtin `@allowDefers();` might be more obvious.

**Comment by paperdave** [comment]
i don't think this should be specific to `if` statements, but just blocks that include defer as the only item.

**Comment by rohlem** [comment]
@paperdave It could in fact be generalized to any sequence of `defer`/`errdefer` at the end of a block (i.e. that aren't followed by any other statement).

**Comment by Cloudef** [comment]
Perhaps also the error could give a  hint how to do what they want
```zig
defer if (cond) later();
```
